### PR TITLE
oq: init at 1.0.2

### PR DESCRIPTION
**Pull request by Br1ght0ne**
###### Motivation for this change
https://github.com/Blacksmoke16/oq/releases/tag/v1.0.2

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/75042)

### CODE REVIEW

**Review by stigtsp — COMMENTED** [review]
Things tested:

- Builds on x86_64-linux (NixOS) OK
- Tested basic usage of  `oq` binary : Fails at parsing xml:
```
echo -e "person:\n name: Sandy\n number: 42\n" | ./oq/bin/oq -i yaml -o xml --xml-root person .person | ./oq/bin/oq -i xml -o json
Invalid input format: 'xml'
Failed to raise an exception: END_OF_STACK
[0x456bd6] ???
[0x4077fb] __crystal_raise +43
[0x420525] ???
[0x40a2b3] main +147
[0x7f225f1b8b8e] __libc_start_main +238
[0x40483a] _start +42
[0x0] ???

```

Result of `nix-review pr 75042` [1](https://github.com/Mic92/nix-review)
<details>
  <summary>1 package were build:</summary>

  - oq
</details>

**Comment by filalex77** [comment]
@stigtsp I think that XML deserialization is implemented in https://github.com/Blacksmoke16/oq/commit/960e10e771a3232c9ce8c935fe6ead2f4e922db5, which was later than the last release. Help text doesn't reflect the actual capabilities at the moment of the release.

When the new release comes out, I'll update it.

**Review by stigtsp — COMMENTED** [review]
`oq` fails to build on darwin. Not sure how critical this is, One option is setting `meta.platform =  plaforms.linux`.

Tested on a local darwin build box, which gave the following output

```

$ nix build --no-link --keep-going --max-jobs 4 --option build-use-sandbox true -f /Users/user/.cache/nix-review/pr-75042/build.nix
builder for '/nix/store/ixr2816qj3jinmx817i14affjqcf6awz-crystal-0.30.1.drv' failed with exit code 2; last 10 log lines:
  Using /nix/store/g8w64ax6s3lzv1kpwq38r1cgpayxqh7k-llvm-6.0.1/bin/llvm-config [version=6.0.1]
  clang++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/nix/store/g8w64ax6s3lzv1kpwq38r1cgpayxqh7k-llvm-6.0.1/include  -stdlib=libc++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wcovered-switch-default -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -O3 -DNDEBUG  -fno-exceptions -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
  ./bin/crystal docs -b https://crystal-lang.org/api/latest src/docs_main.cr
  make: Nothing to be done for 'docs'.
  clang -fPIC    -c -o src/ext/sigfault.o src/ext/sigfault.c
  ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
  CRYSTAL_CONFIG_PATH="/private/var/folders/g1/y968ywwx2s9c016g26_dbpcm0000gn/T/nix-build-crystal-0.30.1.drv-0/source/src" CRYSTAL_CONFIG_LIBRARY_PATH="/nix/store/ikhr2c3zqdjljry2y9rhwc0k1z8k1iy5-crystal-0.30.1/lib/crystal:/nix/store/pr6crirn4s0bamk4ygh8mddz8kb95dw0-crystal-binary-0.29.0/embedded/lib" CRYSTAL_CONFIG_BUILD_COMMIT="" ./bin/crystal build --release --single-module -D preview_overflow -D compiler_rt  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
  Assertion failed: (cast<DISubprogram>(Scope)->describes(MF->getFunction())), function getOrCreateRegularScope, file /var/cache/omnibus/src/llvm/llvm-3.9.1.src/lib/CodeGen/LexicalScopes.cpp, line 159.
  /nix/store/pr6crirn4s0bamk4ygh8mddz8kb95dw0-crystal-binary-0.29.0/bin/crystal: line 103: 17491 Abort trap: 6           "$INSTALL_DIR/embedded/bin/crystal" "$@"
  make: *** [Makefile:132: .build/crystal] Error 134
cannot build derivation '/nix/store/4qgjmgazkd159wd4ihjfnhybxn71ich2-crystal-0.31.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/v9kycmclaglmdi9i375qc6i5w59s2a2k-oq-0.2.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qncdf8xv0xnr1pwf6v0ksfn2f90cra12-env.drv': 1 dependencies couldn't be built
[3 built (1 failed), 160 copied (1734.3 MiB), 342.9 MiB DL]
error: build of '/nix/store/qncdf8xv0xnr1pwf6v0ksfn2f90cra12-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/75042
1 package failed to build:
oq

```

File: pkgs/development/tools/oq/default.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  version = "0.2.1";
+
+  src = fetchFromGitHub {
+    owner = "Blacksmoke16";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0wc1wdvqybgybq09dr78b93mdyvr7avya5w8kffbv8p4i9yyrfn4";
+  };
+
+  buildInputs = [ libxml2 ];
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix]: Consider adding `jq` to buildInputs or propagatedBuildInputs 

(I'm not too sure about how $PATH is handled in crystal packaging)

Comment (stigtsp) [reply pkgs/development/tools/oq/default.nix]: It fails on environments that don't have `jq` in the $PATH. (Mine did when testing)

```
$ ./results/oq/bin/oq 
oq error: execvp (jq): No such file or directory: No such file or directory
```

Comment (filalex77) [reply pkgs/development/tools/oq/default.nix]: Sure, didn't notice that. Thanks!

Comment (stigtsp) [reply pkgs/development/tools/oq/default.nix]: The convo is marked as resolved, but can't see any fix for this problem.

File: pkgs/development/tools/oq/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{ lib, fetchFromGitHub, crystal, jq, libxml2 }:
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix]: Add `makeWrapper`:

`{ lib, fetchFromGitHub, crystal, jq, libxml2, makeWrapper }:`

File: pkgs/development/tools/oq/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  pname = "oq";
+  version = "0.2.1";
+
+  src = fetchFromGitHub {
+    owner = "Blacksmoke16";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0wc1wdvqybgybq09dr78b93mdyvr7avya5w8kffbv8p4i9yyrfn4";
+  };
+
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix]: Add `makeWrapper` as nativeBuildInputs:

`nativeBuildInputs = [ makeWrapper ];`

File: pkgs/development/tools/oq/default.nix [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  buildInputs = [ jq libxml2 ];
+
+  crystalBinaries.oq.src = "src/oq_cli.cr";
+
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix]: Use `wrapProgram` in postInstall to create a wrapper that includes the $PATH to jq:

```
postInstall = ''
  wrapProgram "$out/bin/oq" \
    --prefix PATH : "${lib.makeBinPath [ jq ]}"
'';
```

**Review by stigtsp — CHANGES_REQUESTED** [review]
It seems a wrapper is needed for `oq` to find `jq` when it's not in the environments path. I've added code suggestions.

**Comment by filalex77** [comment]
@stigtsp Thanks a lot for your help! I'm still learning these things.

**Review by stigtsp — APPROVED** [review]
Looks good to me

Result of `nix-review pr 75042` [1](https://github.com/Mic92/nix-review)
<details>
  <summary>1 package were build:</summary>

  - oq
</details>

**Comment by filalex77** [comment]
@stigtsp Is anything impeding this PR from being merged?

**Comment by stigtsp** [comment]
> @stigtsp Is anything impeding this PR from being merged?

Not AFAIK, need someone with merge access to review.

**Comment by stigtsp** [comment]
@filalex77  Version 1.0.0 has been released

File: pkgs/development/tools/oq/default.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{ lib, fetchFromGitHub, crystal, jq, libxml2, makeWrapper }:
+
+crystal.buildCrystalPackage rec {
+  pname = "oq";
+  version = "0.2.1";
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix:5]:
```suggestion
  version = "1.0.2";
```

File: pkgs/development/tools/oq/default.nix [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  src = fetchFromGitHub {
+    owner = "Blacksmoke16";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0wc1wdvqybgybq09dr78b93mdyvr7avya5w8kffbv8p4i9yyrfn4";
```
Comment (stigtsp) [inline pkgs/development/tools/oq/default.nix:11]:
```suggestion
    sha256 = "0sf6rb5b6g7gzyq11l5868p3a1s5z8432swlpv457bfbbnbg6j6q";
```

**Review by stigtsp — COMMENTED** [review]
Update to 1.0.2 tested with:

```shell
$ echo -e "person:\n name: Sandy\n number: 42\n" | ./oq -i yaml -o xml --xml-root person .person | ./oq -i xml -o json | ./oq .person.number
"42"
```

Also needs a rebase

**Comment by filalex77** [comment]
@GrahamcOfBorg eval

**Review by stigtsp — APPROVED** [review]
Looks good to me

Result of `nixpkgs-review pr 75042` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>

  - oq
</details>

**Comment by stigtsp** [comment]
Cc @etu

**Comment by etu** [comment]
Looks reasonable enough, thanks @stigtsp :slightly_smiling_face:

**Comment by stigtsp** [comment]
@filalex77 thx for the pr

**Pull request merged**